### PR TITLE
Automating the creation of the uBlock Origin version of the ClearURLs list

### PR DESCRIPTION
**Pull request by iam-py-test**
It creates a PR, but I can have it just push to the repo
CCing @rusty-snake @DandelionSprout

### CODE REVIEW

**Comment by iam-py-test** [comment]
What just happened?

File: scripts/compile.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+) -> None:
+    for rule in rules:
+        if is_regex(rule):
+            endrules.write(regex_fromat.format(rule, url_pattern) + "\n")
+        else:
+            endrules.write(plain_format.format(rule, url_pattern) + "\n")
+
+def getrules():
+  import requests
+  return requests.get("https://raw.githubusercontent.com/ClearURLs/Rules/master/data.min.json").text
```
Comment (rusty-snake) [inline scripts/compile.py:107]:
```suggestion
def getrules() -> str:
  RULES = "https://raw.githubusercontent.com/ClearURLs/Rules/master/data.min.json"
  return requests.get(RULES).text


```

File: scripts/compile.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+def main() -> int:
+    global endrules
+    data_min_json = json.loads(getrules())
+    endrules = open("uBO list extensions/clear_urls_uboified.txt","w")
```
Comment (rusty-snake) [inline scripts/compile.py:111]:
```suggestion
    endrules = open("uBO list extensions/clear_urls_uboified.txt", "w")
```

File: scripts/compile.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# this code has been modified by https://github.com/iam-py-test, as to improve compatibility and fix issues
+
+import json
+import sys
```
Comment (rusty-snake) [inline scripts/compile.py:25]:
```suggestion
import json
import sys

import requests
```

File: scripts/compile.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+endrules = None
+
```
Comment (rusty-snake) [inline scripts/compile.py:28]:
```suggestion


```

**Review by rusty-snake — COMMENTED** [review]
Looks correct (can't say to much about the workflow). I'm not sure how strict you are about code styling, so fell free to ignore the suggestions below.

File: scripts/compile.py (hunk 1 above)
Comment (iam-py-test) [reply scripts/compile.py:107]: Ah. I'm not great at code styling, but I did once read about something once, so I will push this

File: scripts/compile.py (hunk 2 above)
Comment (iam-py-test) [reply scripts/compile.py:113]: Good catch.

File: scripts/compile.py (hunk 3 above)
Comment (iam-py-test) [reply scripts/compile.py:25]: Why the big space?
I do agree that it should go at the top

File: scripts/compile.py (hunk 4 above)
Comment (iam-py-test) [reply scripts/compile.py:28]: Not sure. I guess...

**Comment by iam-py-test** [comment]
I assume that you are ok with me copying your code?

File: scripts/compile.py (hunk 3 above)
Comment (rusty-snake) [reply scripts/compile.py:25]: > Why the big space?

[PEP 8](https://www.python.org/dev/peps/pep-0008/) says

> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.
>
> Imports should be grouped in the following order:
>
> 1. *Standard library imports.*
> 2. *Related third party imports.*
> 3. Local application/library specific imports.
>
> *You should put a blank line between each group of imports.*

**Comment by rusty-snake** [comment]
> I assume that you are ok with me copying your code?

Absolutely, https://github.com/arkenfox/user.js/issues/1263#issuecomment-949932313:

> @crssi Currently I'm not planing to publish it anywhere (and will delete the gist sooner or later). But if anyone uses it as a start, fixes the last bugs, implements exceptions and sets up a github actions which periodically (e.g. weekly) runs it ...

File: scripts/compile.py (hunk 3 above)
Comment (rusty-snake) [reply scripts/compile.py:25]: FWIW: If you want to automate such styling things you can use [isort](https://pypi.org/project/isort/) for imports and [black](https://pypi.org/project/black/) for everything else. Or laxer formatter like [yapf](https://pypi.org/project/yapf/) or [autopep8](https://pypi.org/project/autopep8/). Sadly, they are not magic tools that turn any code into good code.

Comment (iam-py-test) [reply scripts/compile.py:25]: I partially taught myself Python, so I guess I never read the standards 
Thanks

**Comment by DandelionSprout** [comment]
Okay, so I feel like this is proceeding *way* too fast, to the point I feel I'm beginning to lose control of the situation.

Off the top of my head, one option I'd be more comfortable with, would be if it was possible to place the three files in the same folder, so their filepaths become e.g. `ClearURLs Converter/generate_clearurlsubo.yml`, `ClearURLs Converter/compile.py`, and `ClearURLs Converter/clear_urls_uboified.txt`.

I also think the metadata of `clear_urls_uboified.txt` would need to be re-instated in some form.

File: scripts/compile.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+# SOFTWARE.
+
+# this code has been modified by https://github.com/iam-py-test, as to improve compatibility and fix issues
+
+import json
+import sys
+
+import requests
+
+endrules = None
```
Comment (rusty-snake) [inline scripts/compile.py:29]:
```suggestion
HEAD = """\
! Title: ClearURLs for uBo - List extension
! Homepage: https://github.com/DandelionSprout/adfilt/discussions/163
! Last updated: {date}
! Note: This was forked from https://gist.github.com/rusty-snake/5cd83a87d680ecbd03e79a1a06758207, which is based off of https://github.com/ClearURLs/Rules. The maintainers of Adfilt (DandelionSprout and iam-py-test, and contributors) have made some modifications as to keep it up-to-date with the source and to fix issues
"""

endrules = None
```

File: scripts/compile.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# this code has been modified by https://github.com/iam-py-test, as to improve compatibility and fix issues
+
+import json
+import sys
+
+import requests
```
Comment (rusty-snake) [inline scripts/compile.py:27]:
```suggestion
import json
import sys
from datetime import date

import requests
```

File: scripts/compile.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+def getrules() -> str:
+  RULES = "https://raw.githubusercontent.com/ClearURLs/Rules/master/data.min.json"
+  return requests.get(RULES).text
+
+
+def main() -> int:
+    global endrules
+    data_min_json = json.loads(getrules())
+    endrules = open("uBO list extensions/clear_urls_uboified.txt", "w")
```
Comment (rusty-snake) [inline scripts/compile.py:116]:
```suggestion
    endrules = open("uBO list extensions/clear_urls_uboified.txt", "w")
    endrules.write(HEAD.format(date.today().strftime("%d/%m/%Y")))
```

**Review by rusty-snake — COMMENTED** [review]
Untested try to generated the head.

----

> `ClearURLs Converter/generate_clearurlsubo.yml`

The workflow must be in `.github/workflows`.

**Comment by iam-py-test** [comment]
> Okay, so I feel like this is proceeding _way_ too fast, to the point I feel I'm beginning to lose control of the situation.
> 
> Off the top of my head, one option I'd be more comfortable with, would be if it was possible to place the three files in the same folder, so their filepaths become e.g. `ClearURLs Converter/generate_clearurlsubo.yml`, `ClearURLs Converter/compile.py`, and `ClearURLs Converter/clear_urls_uboified.txt`.
> 

The YML can’t be in the same folder, it has to be in the .github/workflows folder. I could put both the script and output in a new folder, but that would break it for anyone who used the list. Maybe the script should go in the folder with the output like the alternate versions of the antimalware list  

> I also think the metadata of `clear_urls_uboified.txt` would need to be re-instated in some form.

I’m working on that, it just was late yesterday so I had to stop & continue today

File: scripts/compile.py (hunk 7 above)
Comment (iam-py-test) [reply scripts/compile.py:116]: Good idea, but doesn’t it need an ! in front of it

**Comment by DandelionSprout** [comment]
>The YML can’t be in the same folder, it has to be in the .github/workflows folder.

Fair point. I thus give my approval to keeping it in `.github/workflows`.
————————
>I could put both the script and output in a new folder, but that would break it for anyone who used the list

Since the list is currently not listed on FilterLists.com, I presume we can change its folder and thus URL this time around.

**Comment by rusty-snake** [comment]
> I could put both the script and output in a new folder, but that would break it for anyone who used the list

https://help.eyeo.com/adblockplus/how-to-write-filters

> `! Redirect: http://example.com/list.txt` - This comment indicates that the filter list has moved to a new download address. Adblock Plus ignores any file content beyond that comment and immediately tries downloading from the new address. In case of success, the address of the filter list is updated in the settings. This comment is ignored if the new address is the same as the current address, meaning that it can be used to enforce the "canonical" address of the filter list.

Does anyone know if this this is supported by µBlock origin?

**Comment by iam-py-test** [comment]
> > The YML can’t be in the same folder, it has to be in the .github/workflows folder.
> 
> Fair point. I thus give my approval to keeping it in `.github/workflows`. ————————
> 
> > I could put both the script and output in a new folder, but that would break it for anyone who used the list
> 
> Since the list is currently not listed on FilterLists.com, I presume we can change its folder and thus URL this time around.

To my knowledge, the only people who have subscribed to it is me, @kah0922, and maybe @rusty-snake 
So do you think the folder should be called something like _ClearURLs for uBo_?

File: scripts/compile.py (hunk 7 above)
Comment (rusty-snake) [reply scripts/compile.py:116]: Where do you mean to add a `!`? Anyway, I tested this locally and it works.

**Comment by iam-py-test** [comment]
> 
> 
> > I could put both the script and output in a new folder, but that would break it for anyone who used the list
> 
> https://help.eyeo.com/adblockplus/how-to-write-filters
> 
> > `! Redirect: http://example.com/list.txt` - This comment indicates that the filter list has moved to a new download address. Adblock Plus ignores any file content beyond that comment and immediately tries downloading from the new address. In case of success, the address of the filter list is updated in the settings. This comment is ignored if the new address is the same as the current address, meaning that it can be used to enforce the "canonical" address of the filter list.
> 
> Does anyone know if this this is supported by µBlock origin?

No

**Comment by rusty-snake** [comment]
> We could make a post on #163 though

If people are aware of the list, they probably read in #163. So yes.

File: scripts/compile.py (hunk 5 above)
Comment (iam-py-test) [reply scripts/compile.py:29]: Ah, good idea. Now I understand...

File: scripts/compile.py (hunk 7 above)
Comment (iam-py-test) [reply scripts/compile.py:117]: Wait. I'm just confused. Your right

**Comment by iam-py-test** [comment]
I modified it to create a PR but it seems GH had a mind of its own and reverted my changes
@DandelionSprout would you rather a PR or just a direct commit

**Comment by iam-py-test** [comment]
Ok. I need help
The fork has gone crazy & is pushing commits to different branches randomly. I think it is because this repo uses _master_ as the main branch, while GitHub gives new repos the default branch of main

**Comment by DandelionSprout** [comment]
>would you rather a PR or just a direct commit

PR, mostly because I need to handle vacuum cleaning and similar cleaning before I can tend to this in detail.

**Comment by iam-py-test** [comment]
> 
> 
> > would you rather a PR or just a direct commit
> 
> PR, mostly because I need to handle vacuum cleaning and similar cleaning before I can tend to this in detail.

Ok. I think I need to close this PR and open a new one, because GH is making it impossible to use this one

**Comment by iam-py-test** [comment]
Ok. Closing this in favor of #327

**Pull request closed without merging**